File: programmers-coding-test/Java/Level1/옹알이.java
class Solution {

    final String[] pronounce = {"aya", "ye", "woo", "ma"};

    public int solution(String[] babbling) {
        int answer = 0;

        for (String b : babbling) {
            String before = "";
            boolean canPronounce = true;

            while (!b.isEmpty()) {
                boolean can = false;
                for (String p : pronounce) {
                    if (!p.equals(before) && b.indexOf(p) == 0) {
                        b = b.substring(p.length());
                        before = p;
                        can = true;
                        break;
                    }
                }
                if (!can) {
                    canPronounce = false;
                    break;
                }
            }
            if (canPronounce)
                answer++;
        }

        return answer;
    }
}

File: programmers/C++/Level1/옹알이.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

unordered_set<string> pronounce{ "aya", "ye", "woo", "ma" };

bool canPronounce(string str, string before) {
    if (str.empty())
        return true;

    for (string p : pronounce) 
        if (p != before && str.find(p) == 0) 
            return canPronounce(str.substr(p.length()), p);

    return false;
}

int solution(vector<string> babbling) {
    int answer = 0;

    for (string b : babbling) 
        if (canPronounce(b, ""))
            answer++;

    return answer;
}